New resource - `azurerm_palo_alto_next_generation_firewall_virtual_*_strata_cloud_manager`

QixiaLu opened the pull request:
<!--  All Submissions -->

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Add new resources to support palo alto network firewall strata cloud manager addition:
`azurerm_palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager`
`azurerm_palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager`.

API: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/paloaltonetworks/resource-manager/PaloAltoNetworks.Cloudngfw/stable/2025-05-23/PaloAltoNetworks.Cloudngfw.json

MSFT doc: https://learn.microsoft.com/en-us/rest/api/cloudngfw/firewalls/create-or-update?view=rest-cloudngfw-2025-05-23&tabs=HTTP#ipaddress


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager` - New resource support for SCM addition
* `azurerm_palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager` - New resource support for SCM addition


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #30095 

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

JoeRistine-Synapse commented:
Hello, is there a timeline for this?

QixiaLu commented:
> Hello, is there a timeline for this?

Hi @JoeRistine-Synapse , thanks for reaching out. This PR is currently blocked by an issue in Azure. We noticed that `plan_id` is not able to be updated. The service team is currently debugging on this. Once it's fixed or we find a workaround, we'll continue working on this resource.

QixiaLu commented:
<img width="1341" height="610" alt="Screenshot 2025-10-23 162255" src="https://github.com/user-attachments/assets/ffc2ad62-f473-4792-a311-986713566c7a" />

<img width="1180" height="343" alt="Screenshot 2025-10-23 162147" src="https://github.com/user-attachments/assets/d4cf1688-57d3-491e-9146-db693c2025c3" />

basic / import / complete tests pass. There's an issue with updating `plan_id` in the Azure API. Will fix the test after the upstream issue is fixed.

liuwuliuyun reviewed (commented):
Overall looks good, just some minor changes I marked in comments.

QixiaLu commented on internal/services/paloalto/helpers.go:
```diff
@@ -0,0 +1,49 @@
+
+	output, err := identity.FlattenUserAssignedMapToModel(&tmp)
+	if err != nil {
+		return nil, fmt.Errorf("expanding `identity`: %+v", err)
```
updated, thanks!

liuwuliuyun reviewed (approved):
LGTM, Thanks

wuxu92 commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,331 @@
+				if !response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+				}
+			}
```
we can combine conditions to one:

```suggestion
        if err != nil && !response.WasNotFound(existing.HttpResponse) {
			return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
		}
````

wuxu92 commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,331 @@
+
+				state.MarketplaceOfferId = props.MarketplaceDetails.OfferId
+
+				state.PlanId = props.PlanData.PlanId
```
```suggestion
				state.MarketplaceOfferId = props.MarketplaceDetails.OfferId
				state.PlanId = props.PlanData.PlanId
```

wuxu92 commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+
+				state.MarketplaceOfferId = props.MarketplaceDetails.OfferId
+
+				state.PlanId = props.PlanData.PlanId
```
```suggestion
				props := model.Properties
				state.Location = location.Normalize(model.Location)
				state.DNSSettings = schema.FlattenDNSSettings(props.DnsSettings)
				state.FrontEnd = schema.FlattenDestinationNAT(props.FrontEndSettings)
				state.NetworkProfile = schema.FlattenNetworkProfileVnet(props.NetworkProfile)
				state.MarketplaceOfferId = props.MarketplaceDetails.OfferId
				state.PlanId = props.PlanData.PlanId
```

sreallymatt commented on internal/services/paloalto/helpers.go:
```diff
@@ -0,0 +1,49 @@
+// tracked on https://github.com/Azure/azure-rest-api-specs/issues/38348
+// identity defined both systemAssigned and userAssigned Identity type in Swagger but only support userAssigned Identity,
+// so add a workaround to convert type here.
+func expandPaloAltoLegacyToUserAssignedIdentity(input []identity.ModelUserAssigned) (*identity.LegacySystemAndUserAssignedMap, error) {
```
Instead of adding additional functions to handle this, could you add a workaround to Pandora so we can fix this data issue at the SDK level? 

You can use [this PR](https://github.com/hashicorp/pandora/pull/5060/files) as an example of what that would look like

QixiaLu replied:
Added a workaround in pandora, could I get a review on this https://github.com/hashicorp/pandora/pull/5077? After the SDK is upgraded, I'll remove the identity workaround here. Thanks!

sreallymatt replied:
New version available: `v0.20251126.1220923`, while we generally update `go-azure-sdk` separately, I think in this case it's fine to include it in this PR since the only changes are specifically for these new resources. (The changes in the `network` package are a known issue with the SDK generator, they always get added, then removed in the next cycle)

QixiaLu replied:
Thanks Matt! Let me update `go-azure-sdk` in this PR and remove current workaround

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+				if response.WasNotFound(existing.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading %s: %+v", *id, err)
```
We prefer `retrieving` over `reading` in error messages on GET requests. For consistency, could you update this?

```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+				if response.WasNotFound(existing.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading %s: %+v", *id, err)
```
```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+				if err := validate.ValidateStrataCloudManagerTenantNameExists(ctx, metadata.Client, metadata.Client.Account.SubscriptionId, model.StrataCloudManagerTenantName); err != nil {
+					return fmt.Errorf("validating strata_cloud_manager_tenant_name: %+v", err)
+				}
+				props.StrataCloudManagerConfig.CloudManagerName = model.StrataCloudManagerTenantName
```
While unlikely to happen being that `strata_cloud_manager_tenant_name` is required, we should still add a check here on `StrataCloudManagerConfig` to ensure this can't panic in the event Azure returns a nil value

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+func (r NextGenerationFirewallVNetStrataCloudManagerResource) Arguments() map[string]*pluginsdk.Schema {
+	args := map[string]*pluginsdk.Schema{
```
```suggestion
	return map[string]*pluginsdk.Schema{
```

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,319 @@
+				if response.WasNotFound(existing.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading %s: %+v", *id, err)
```
```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

sreallymatt commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,319 @@
+				if response.WasNotFound(existing.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading %s: %+v", *id, err)
```
```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

sreallymatt commented on website/docs/r/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
```
```suggestion
The `timeouts` block allows you to specify [timeouts](https://developer.hashicorp.com/terraform/language/resources/configure#define-operation-timeouts) for certain actions:
```

sreallymatt commented on website/docs/r/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager.html.markdown:
```diff
@@ -0,0 +1,259 @@
+
+* `location` - (Required) The Azure Region where the Palo Alto Next Generation Firewall Virtual Network Strata Cloud Manager should exist. Changing this forces a new Palo Alto Next Generation Firewall Virtual Network Strata Cloud Manager to be created.
+
+* `name` - (Required) The name which should be used for this Palo Alto Next Generation Firewall Virtual Network Strata Cloud Manager. Changing this forces a new Palo Alto Next Generation Firewall Virtual Network Strata Cloud Manager to be created.
```
Same as my comment on the vhub resource, please flip the order of these 2 args

sreallymatt reviewed (changes requested):
Thanks @QixiaLu, I've left some comments inline

QixiaLu commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+
+			if err := validate.ValidateStrataCloudManagerTenantNameExists(ctx, metadata.Client, metadata.Client.Account.SubscriptionId, model.StrataCloudManagerTenantName); err != nil {
+				return fmt.Errorf("validating strata_cloud_manager_tenant_name: %+v", err)
+			}
```
Moved to CustomizeDiff in vhub and vnet resources, thanks!

QixiaLu commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,325 @@
+}
+
+func (r NextGenerationFirewallVHubStrataCloudManagerResource) Arguments() map[string]*pluginsdk.Schema {
+	args := map[string]*pluginsdk.Schema{
```
Fixed this in vhub and vnet

QixiaLu commented on website/docs/r/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+* `location` - (Required) The Azure Region where the Palo Alto Next Generation Firewall Virtual Hub Strata Cloud Manager should exist. Changing this forces a new Palo Alto Next Generation Firewall Virtual Hub Strata Cloud Manager to be created.
+
+* `name` - (Required) The name which should be used for this Palo Alto Next Generation Firewall Virtual Hub Strata Cloud Manager. Changing this forces a new Palo Alto Next Generation Firewall Virtual Hub Strata Cloud Manager to be created.
```
Updated this in both vhub and vnet, cheers!

QixiaLu commented on website/docs/r/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager.html.markdown:
```diff
@@ -0,0 +1,259 @@
+
+* `dns_servers` - (Optional) A list of DNS servers to use. Conflicts with `dns_settings[0].use_azure_dns`.
+
+* `use_azure_dns` - (Optional) Should the Firewall use Azure Supplied DNS servers. Conflicts with `dns_settings[0].dns_servers`. Defaults to `false`.
```
Updated to `dns_settings[0].<field>`. cheers!

QixiaLu commented:
Hi @sreallymatt , I have addressed your comments above. Here is the test results after the update. 
<img width="1564" height="315" alt="Screenshot 2025-11-28 153621" src="https://github.com/user-attachments/assets/3ab9b7f6-d397-4795-847e-d8e5f4628121" />
<img width="1712" height="311" alt="Screenshot 2025-11-28 103108" src="https://github.com/user-attachments/assets/23790486-99b2-4af8-a05c-f8fb8741fc9d" />


Sorry that I have to rebase the PR which pollutes this PR commit messages, because this branch was from a quite old Main history and Git not able to find common history commit :(

sreallymatt reviewed (changes requested):
Thanks @QixiaLu, 2 minor changes to make then this should be good to go!

QixiaLu commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_hub_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,342 @@
+
+			if metadata.ResourceData.HasChange("strata_cloud_manager_tenant_name") {
+				if props.StrataCloudManagerConfig == nil {
+					return fmt.Errorf("updating %s: StrataCloudManagerConfig was nil", *id)
```
Thanks Matt! Updated to initialisation rather than panic

QixiaLu commented on internal/services/paloalto/palo_alto_next_generation_firewall_virtual_network_strata_cloud_manager_resource.go:
```diff
@@ -0,0 +1,336 @@
+
+			if metadata.ResourceData.HasChange("strata_cloud_manager_tenant_name") {
+				if props.StrataCloudManagerConfig == nil {
+					return fmt.Errorf("updating %s: StrataCloudManagerConfig was nil", *id)
```
Updated, thank you!
